Turn on UseClangFunctionTypes by default.

varungandhi-apple opened the pull request:
~~Currently, some tests are failing, working on fixing those.~~

varungandhi-apple committed:
Enable UseClangFunctionTypes by default.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test source compatibility

rjmccall commented on lib/SILGen/SILGenThunk.cpp:
```diff
@@ -159,10 +161,32 @@ SILGenModule::getOrCreateForeignAsyncCompletionHandlerImplFunction(
+      llvm::copy(fnProtoTy->getParamTypes(), std::back_inserter(implClangArgs));
+      auto newFnTy = clangCtx.getFunctionType(
+          fnProtoTy->getReturnType(), implClangArgs,
+          fnProtoTy->getExtProtoInfo());
```
You can't just use the `ExtProtoInfo` here because any `ExtParameterInfo`s need to be adjusted by one to compensate for the new argument.

Also, you should make a utility function somewhere to do this.

varungandhi-apple replied:
Fixed.

rjmccall commented on lib/SILGen/SILGenThunk.cpp:
```diff
@@ -159,10 +161,32 @@ SILGenModule::getOrCreateForeignAsyncCompletionHandlerImplFunction(
+
+  const clang::Type *clangType = blockType->getClangTypeInfo().getType();
+  const clang::Type *newClangType = nullptr;
+  if (auto blockPtrTy = dyn_cast_or_null<clang::BlockPointerType>(clangType)) {
```
Can this be null?  Can the cast fail?

varungandhi-apple replied:
It can be null because `UseClangFunctionTypes` might be turned off (and this fix is independent of that). It can't fail though.

varungandhi-apple replied:
Fixed.

rjmccall commented on lib/SILGen/SILGenThunk.cpp:
```diff
@@ -159,10 +161,32 @@ SILGenModule::getOrCreateForeignAsyncCompletionHandlerImplFunction(
+          fnProtoTy->getExtProtoInfo());
+      newClangType = clangCtx.getPointerType(newFnTy).getTypePtr();
+    } else {
+      llvm::report_fatal_error("Unexpected FunctionNoProtoType!");
```
Did you test that this can't happen if there's a completion handler of type e.g. `void(^)()`?

varungandhi-apple replied:
I've handled the case for now, but non-prototype completion handlers don't seem to be working. Will investigate/file a bug report.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please test source compatibility

varungandhi-apple commented:
@swift-ci please smoke test macOS

varungandhi-apple commented:
@swift-ci please clean test macOS

varungandhi-apple commented:
@swift-ci please clean smoke test macOS

varungandhi-apple commented:
There is a Clang type mismatch when building GRDB (failure in source compat suite) looking at that first before the SourceKitLSP issue.

rjmccall reviewed (approved):
LGTM

varungandhi-apple committed:
[test] Add crash test for attributed Clang types.

varungandhi-apple committed:
[SILGen] Fix crash when compiling attributed Clang function types.

Before this change, @convention(c|block) function types with non-trivial
Clang types would not be compiled correctly, because the AbstractionPattern
would sometimes be of Kind::Type instead of Kind::ClangType. We normalize
the AbstractionPattern right before lowering parameters and results to make
sure that any attributes on the stored Clang type are respected.

varungandhi-apple commented:
There are a couple of test failures but everything else seems to work OK.

- [ ] `SILGen/objc_blocks_bridging.swift`
- [ ] `SILGen/objc_bridging.swift`

I will look at those next week.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
There is something about this mutation of pattern that makes me a little bit uneasy. On one hand, `hasNonDerivableClangType()` will return true only for Clang types coming from C, so I think it should be OK to change the abstraction pattern. OTOH, looks like if you do this indiscriminately, then something like `@convention(block) (String) -> ()` would get lowered to `@convention(block) (Optional<NSString>) -> ()` when the pattern is of kind `ClangType`, but it would be lowered to `@convention(block) (NSString) -> ()` if the pattern is of kind `Type`.

This had led to the failure in tests like `SILGen/objc_bridging.swift` noted earlier (we didn't have the `hasNonDerivableClangType()` check).
